### PR TITLE
MRG: Fixes for 1 channel GDF, millivolt support

### PR DESCRIPTION
**Pull request by abramhindle**
#### What does this implement/fix?
I have a 1 channel GDF file that failed to load because the raw_headers were encapsulated in lists.
I added a check for this condition to enable the file to load -- with runtime warnings. 
I fixed 1 runtime warning by adding support for millivots in GDF files.
The tests that I added create the problematic 1 channel GDF ECG file---my own measurement of myself, don't judge---and then proceed to load the file and ensure it can check the sampling rate and that it loaded the data sanely.

#### Additional information
The test file has an embedded file within it so it is not path sensitive and can always make its testing environment.
There is still 1 runtime warning regarding the filters used. I have not investigated heavily.

In summary:
* I fixed 1 bug
* I added tests with test data.

### CODE REVIEW

File: mne/io/edf/tests/test_one_channel_gdf.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    # pytest.raises(AssertionError, _assert_line_endings, tempdir)
+    # now check mne
+    ecg = mne.io.read_raw_gdf(fname, preload=True)
+    assert ecg['ECG'][0].shape == (1, 4500)
+    assert 150.0 == ecg.info['sfreq']
+
+
+def one_channel_gdf():
+    ''' This is an embedded 1 channel ECG '''
+    b64data = """R0RGIDIuMTAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
```
Comment (larsoner) [inline mne/io/edf/tests/test_one_channel_gdf.py]: It would probably be better to have this in a separate file, esp. in the `mne-testing-data` repo (see for instructions)

https://github.com/mne-tools/mne-testing-data

File: mne/io/edf/tests/test_one_channel_gdf.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+from os import path as op
+
+import pytest
+import mne
+import base64
+
+from mne.utils import run_tests_if_main, _TempDir
+
+# the file in question doesn't have highpass and lowpass
+@pytest.mark.filterwarnings('ignore::RuntimeWarning')
```
Comment (larsoner) [inline mne/io/edf/tests/test_one_channel_gdf.py]: Better to do this below:
```
with pytest.warns(RuntimeWarning, match='highpass'):
    ecg = mne.io.read_raw_gdf(fname, preload=True)
```

File: mne/io/edf/edf.py [hunk 3]
```diff
@@ -302,6 +302,11 @@ def _read_segment_file(data, idx, fi, start, stop, raw_extras, chs, filenames):
     buf_len = int(raw_extras['max_samp'])
     dtype = raw_extras['dtype_np']
     dtype_byte = raw_extras['dtype_byte']
+    if isinstance(dtype, list):  # depending on the type of GDF raw_extras
+        dtype = dtype[0]         # wraps things in lists?
+    dtype_byte = raw_extras['dtype_byte']
+    if isinstance(dtype_byte, list):  # depending on the type of GDF raw_extras
+        dtype_byte = dtype_byte[0]   # wraps things in lists?
```
Comment (larsoner) [inline mne/io/edf/edf.py]: I am surprised that this is necessary, in theory these things should nicely generalize even down to the 1-channel level. Can you paste the traceback you get without this change? There might be a cleaner fix somehow.

Comment (abramhindle) [reply mne/io/edf/edf.py]: >       n_per = max(10 * 1024 * 1024 // (ch_offsets[-1] * dtype_byte), 1)
    E       TypeError: unsupported operand type(s) for //: 'int' and 'list'

Here's the trace. Fundamentally there's a multiplication/division against a list (dtype_byte) that should have been a number. Then later if dtype is not extracted from the list it tells numpy to make an invalid type. raw_settings has list quoted many of the settings.

We can see below that dtype and dtype_byte are lists and not numpy arrays or anything fun.
```
dtype      = [<class 'numpy.float32'>]
dtype_byte = [4]
```

```
hindle1@st-francis:~/projects/mne-python$ pytest mne/io/edf/tests/test_one_channel_gdf.py 
Test session starts (platform: linux, Python 3.6.8, pytest 5.1.2, pytest-sugar 0.9.2)
rootdir: /home/hindle1/projects/mne-python, inifile: setup.cfg
plugins: cov-2.7.1, mock-1.10.4, timeout-1.3.3, sugar-0.9.2
collecting ... 

―――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_one_channel_gdf ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

    @pytest.mark.filterwarnings('ignore::RuntimeWarning')
    def test_one_channel_gdf():
        """Test line endings of mne-python."""
        tempdir = _TempDir()
        fname = op.join(tempdir, 'single_channel.gdf')
        with open(fname, 'wb') as fid:
            fid.write(one_channel_gdf())
        # pytest.raises(AssertionError, _assert_line_endings, tempdir)
        # now check mne
>       ecg = mne.io.read_raw_gdf(fname, preload=True)

fid        = <_io.BufferedWriter name='/tmp/tmp_mne_tempdir_zwl1y870/single_channel.gdf'>
fname      = '/tmp/tmp_mne_tempdir_zwl1y870/single_channel.gdf'
tempdir    = '/tmp/tmp_mne_tempdir_zwl1y870'

mne/io/edf/tests/test_one_channel_gdf.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
mne/io/edf/edf.py:1362: in read_raw_gdf
    verbose=verbose)
</home/hindle1/projects/mne-python/mne/externals/decorator.py:decorator-gen-157>:2: in __init__
    ???
mne/utils/_logging.py:89: in wrapper
    return function(*args, **kwargs)
mne/io/edf/edf.py:253: in __init__
    verbose=verbose)
</home/hindle1/projects/mne-python/mne/externals/decorator.py:decorator-gen-136>:2: in __init__
    ???
mne/utils/_logging.py:89: in wrapper
    return function(*args, **kwargs)
mne/io/base.py:424: in __init__
    self._preload_data(preload)
</home/hindle1/projects/mne-python/mne/externals/decorator.py:decorator-gen-139>:2: in _preload_data
    ???
mne/utils/_logging.py:89: in wrapper
    return function(*args, **kwargs)
mne/io/base.py:669: in _preload_data
    self._data = self._read_segment(data_buffer=data_buffer)
mne/io/base.py:568: in _read_segment
    cals, mult)
</home/hindle1/projects/mne-python/mne/externals/decorator.py:decorator-gen-158>:2: in _read_segment_file
    ???
mne/utils/_logging.py:89: in wrapper
    return function(*args, **kwargs)
mne/io/edf/edf.py:269: in _read_segment_file
    self._filenames[fi])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

data = array([[0., 0., 0., ..., 0., 0., 0.]]), idx = slice(None, None, None), fi = 0, start = 0, stop = 4500
raw_extras = {'bytes_tot': 4, 'ch_names': ['ECG'], 'data_offset': 512, 'digital_max': array([1.649882]), ...}
chs = [{'cal': 3.30057, 'ch_name': 'ECG', 'coil_type': 1 (FIFFV_COIL_EEG), 'coord_frame': 4 (FIFFV_COORD_HEAD), ...}]
filenames = '/tmp/tmp_mne_tempdir_zwl1y870/single_channel.gdf'

    def _read_segment_file(data, idx, fi, start, stop, raw_extras, chs, filenames):
        """Read a chunk of raw data."""
        from scipy.interpolate import interp1d
    
        n_samps = raw_extras['n_samps']
        buf_len = int(raw_extras['max_samp'])
        dtype = raw_extras['dtype_np']
        dtype_byte = raw_extras['dtype_byte']
        data_offset = raw_extras['data_offset']
        stim_channel = raw_extras['stim_channel']
        orig_sel = raw_extras['sel']
        tal_idx = raw_extras.get('tal_idx', [])
        subtype = raw_extras['subtype']
    
        if np.size(dtype_byte) > 1:
            if len(np.unique(dtype_byte)) > 1:
                warn("Multiple data type not supported")
            dtype = dtype[0]
            dtype_byte = dtype_byte[0]
    
        # gain constructor
        physical_range = np.array([ch['range'] for ch in chs])
        cal = np.array([ch['cal'] for ch in chs])
        cal = np.atleast_2d(physical_range / cal)  # physical / digital
        gains = np.atleast_2d(raw_extras['units'])
    
        # physical dimension in uV
        physical_min = raw_extras['physical_min']
        digital_min = raw_extras['digital_min']
    
        offsets = np.atleast_2d(physical_min - (digital_min * cal)).T
        this_sel = orig_sel[idx]
        if len(tal_idx):
            this_sel = np.concatenate([this_sel, tal_idx])
        tal_data = []
    
        # We could read this one EDF block at a time, which would be this:
        ch_offsets = np.cumsum(np.concatenate([[0], n_samps]), dtype=np.int64)
        block_start_idx, r_lims, d_lims = _blk_read_lims(start, stop, buf_len)
        # But to speed it up, we really need to read multiple blocks at once,
        # Otherwise we can end up with e.g. 18,181 chunks for a 20 MB file!
        # Let's do ~10 MB chunks:
>       n_per = max(10 * 1024 * 1024 // (ch_offsets[-1] * dtype_byte), 1)
E       TypeError: unsupported operand type(s) for //: 'int' and 'list'

block_start_idx = 0
buf_len    = 1
cal        = array([[1.]])
ch_offsets = array([0, 1])
chs        = [{'cal': 3.30057, 'ch_name': 'ECG', 'coil_type': 1 (FIFFV_COIL_EEG), 'coord_frame': 4 (FIFFV_COORD_HEAD), ...}]
d_lims     = array([[   0,    1],
       [   1,    2],
       [   2,    3],
       ...,
       [4497, 4498],
       [4498, 4499],
       [4499, 4500]])
data       = array([[0., 0., 0., ..., 0., 0., 0.]])
data_offset = 512
digital_min = array([-1.650688])
dtype      = [<class 'numpy.float32'>]
dtype_byte = [4]
fi         = 0
filenames  = '/tmp/tmp_mne_tempdir_zwl1y870/single_channel.gdf'
gains      = array([[1]])
idx        = slice(None, None, None)
interp1d   = <class 'scipy.interpolate.interpolate.interp1d'>
n_samps    = array([1], dtype=int32)
offsets    = array([[0.]])
orig_sel   = array([0])
physical_min = array([-1.650688])
physical_range = array([3.30057])
r_lims     = array([[0, 1],
       [0, 1],
       [0, 1],
       ...,
       [0, 1],
       [0, 1],
       [0, 1]])
raw_extras = {'bytes_tot': 4, 'ch_names': ['ECG'], 'data_offset': 512, 'digital_max': array([1.649882]), ...}
start      = 0
stim_channel = None
stop       = 4500
subtype    = 'GDF'
tal_data   = []
tal_idx    = []
this_sel   = array([0])

mne/io/edf/edf.py:339: TypeError
---------------------------------------------------------- Captured stdout call -----------------------------------------------------------
Extracting EDF parameters from /tmp/tmp_mne_tempdir_zwl1y870/single_channel.gdf...
GDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 4499  =      0.000 ...    29.993 secs...

 mne/io/edf/tests/test_one_channel_gdf.py ⨯                                                                                 100% ██████████
--------------------------------- generated xml file: /home/hindle1/projects/mne-python/junit-results.xml ---------------------------------


======================================================== slowest 20 test durations ========================================================
3.76s setup    mne/io/edf/tests/test_one_channel_gdf.py::test_one_channel_gdf
0.07s call     mne/io/edf/tests/test_one_channel_gdf.py::test_one_channel_gdf

(0.00 durations hidden.  Use -vv to show these durations.)
========================================================= short test summary info =========================================================
FAILED mne/io/edf/tests/test_one_channel_gdf.py::test_one_channel_gdf - TypeError: unsupported operand type(s) for //: 'int' and 'list'

Results (25.50s):
       1 failed
         - mne/io/edf/tests/test_one_channel_gdf.py:17 test_one_channel_gdf
```

**Comment by larsoner** [comment]
Thanks @abramhindle
